### PR TITLE
Use src directory for project directory structure

### LINKED ISSUE #512: Consider use of src directory for project directory structure
# Description

There is no pressing issue here, but I would like to suggest the idea and have some discussion of moving to using the `src` directory pattern.

Instead of trying to summarize here why I think this should be the case I will let others do so for me:
- PythonBytes [Episode #22 show notes: "Item 5: Python Package src-ery"](https://pythonbytes.fm/episodes/show/22/pythonpath-considered-harmful)
   - [Full transcript of episode](https://pythonbytes.fm/episodes/transcript/22/pythonpath-considered-harmful)
   - pytest's [recommendation for projects that have a `__init__.py` in the `tests` directory](https://docs.pytest.org/en/latest/goodpractices.html#choosing-a-test-layout-import-rules)
   - Ionel Mărieș's argument in his "[Packaging a python library](https://blog.ionelmc.ro/2014/05/25/python-packaging/)" blog post on the subject
- Hynek Schlawack's "[Testing & Packaging](https://hynek.me/articles/testing-packaging/)" blog post
- The PyPA's [summary of their recommendations given their 2019 Packaging Mini Summit](https://github.com/pypa/packaging.python.org/issues/320#issuecomment-495990983)

This isn't an argument, but more libraries are starting to adopt `src` (though to be fair, many do not). A few I know off the top of my head are:
   - [pytest](https://github.com/pytest-dev/pytest)
   - [pyramid](https://github.com/pylons/pyramid)
   - [hypothesis-jsonschema](https://github.com/Zac-HD/hypothesis-jsonschema)

### PR DESCRIPTION
**Pull request by matthewfeickert**
# Description

Resolves #512 

Instead of trying to summarize here why I think this should be the case I will let others do so for me:
- PythonBytes [Episode #22 show notes: "Item 5: Python Package src-ery"](https://pythonbytes.fm/episodes/show/22/pythonpath-considered-harmful)
   - [Full transcript of episode](https://pythonbytes.fm/episodes/transcript/22/pythonpath-considered-harmful)
   - pytest's [recommendation for projects that have a `__init__.py` in the `tests` directory](https://docs.pytest.org/en/latest/goodpractices.html#choosing-a-test-layout-import-rules)
   - Ionel Mărieș's argument in his "[Packaging a python library](https://blog.ionelmc.ro/2014/05/25/python-packaging/)" blog post on the subject
- Hynek Schlawack's "[Testing & Packaging](https://hynek.me/articles/testing-packaging/)" blog post
- The PyPA's [summary of their recommendations given their 2019 Packaging Mini Summit](https://github.com/pypa/packaging.python.org/issues/320#issuecomment-495990983)

TLDR: Use a `src` directory structure to ensure that the code that is tested is the code that is shipped in the distribution and nothing else.

# Checklist Before Requesting Reviewer

- [x] Tests are passing
- [x] "WIP" removed from the title of the pull request
- [x] Selected an Assignee for the PR to be responsible for the log summary

# Before Merging

For the PR Assignees:

- [x] Summarize commit messages into a comprehensive review of the PR

```
* Use a src directory for project directory structure
   - Ensures that the code that is tested is the code that is shipped
   - c.f.
     https://docs.pytest.org/en/latest/goodpractices.html#choosing-a-test-layout-import-rules
     https://blog.ionelmc.ro/2014/05/25/python-packaging/
* Use MANIFEST.in graft command to simplify manifest
   - c.f.
     https://stackoverflow.com/questions/41661616/what-is-the-graft-command-in-pythons-manifest-in-file
     https://blog.ionelmc.ro/2014/06/25/python-packaging-pitfalls/
* Update schema paths for deployment to docs
* Remove the shebang from setup.py
   - c.f. https://www.python.org/dev/peps/pep-0394/
```

### CODE REVIEW

**Comment by lukasheinrich** [comment]
do we need to update `CONTRIBUTING.md`?

**Comment by matthewfeickert** [comment]
> do we need to update `CONTRIBUTING.md`?

Good question. Let me check.

**Comment by matthewfeickert** [comment]
@lukasheinrich Answer is no:

```
$ git grep "pyhf/" CONTRIBUTING.md
CONTRIBUTING.md:If you have found a bug please report it by filling out the [bug report template](https://github.com/diana-hep/pyhf/issues/new?template=Bug-Report.md&labels=bug&title=Bug+Report+:+Title+Here).
```

File: MANIFEST.in [hunk 1]
```diff
@@ -1,6 +1,2 @@
-include pyhf/schemas/1.0.0/defs.json
-include pyhf/schemas/1.0.0/measurement.json
-include pyhf/schemas/1.0.0/model.json
-include pyhf/schemas/1.0.0/workspace.json
-include pyhf/schemas/HistFactorySchema.dtd
+graft src
```
Comment (matthewfeickert) [inline MANIFEST.in]: @kratsg It would be nice to be able to [use graft to be super explicit on what we're packaging](https://blog.ionelmc.ro/2014/06/25/python-packaging-pitfalls/). I'm worried that I've messed up the schema deployment though. Can you explain a bit more about what how the paths are setup for the deployment? Is my updating the symlink like I've done in this PR enough?

**Comment by matthewfeickert** [comment]
The docs need to be fixed, as there is `failed to import` errors.

**Comment by matthewfeickert** [comment]
Build of the docs can be seen here: https://pyhf-staging.readthedocs.io/en/feature-use-src-directory/

**Comment by matthewfeickert** [comment]
@lukasheinrich @kratsg I'm stopping for a bit to write defense slides, but if you have any thoughts on why there are [Cyclic import errors in `setup.py`](https://www.codefactor.io/repository/github/diana-hep/pyhf/pull/513) (:confused:) I'd be keen to hear them.

File: MANIFEST.in (hunk 1 above)
Comment (kratsg) [reply MANIFEST.in]: `graft src` and `include pyhf/schemas` are incompatible. You do not include pyhf source code in the `MANIFEST.in` which is what you've done. You only want to include the schemas as part of _additional_ files to include.

File: docs/Makefile [hunk 2]
```diff
@@ -55,7 +55,7 @@ html:
 	$(SPHINXBUILD) -b html $(ALLSPHINXOPTS) $(BUILDDIR)/html
 	#mkdir -p $(BUILDDIR)/html/docs && ln -sfn ../_static $(BUILDDIR)/html/docs/_static
 	mkdir -p $(BUILDDIR)/html/docs && rsync -r $(BUILDDIR)/html/_static $(BUILDDIR)/html/docs/
-	rsync -r ../pyhf/schemas/ $(BUILDDIR)/html/schemas
```
Comment (kratsg) [inline docs/Makefile]: the paths are wrong. you're changing `pyhf` -> `src`.

File: docs/schemas [hunk 3]
```diff
@@ -0,0 +1 @@
+../src/pyhf/schemas
```
Comment (kratsg) [inline docs/schemas]: what's this? looks like you made this a symlink, and you shouldn't. symlinks don't work with TCP/IP.

Comment (matthewfeickert) [reply docs/schemas]: I was copying what you did in PR #452 https://github.com/diana-hep/pyhf/pull/452/files#diff-03602e9b79dc3f2cb4d9905f64179dbb. Why did you add it there if it isn't needed?

File: MANIFEST.in (hunk 1 above)
Comment (matthewfeickert) [reply MANIFEST.in]: > You do not include pyhf source code in the `MANIFEST.in`

Sorry, can you elaborate why not? [Pyramid](https://github.com/Pylons/pyramid/blob/30aae5f860e90126e45b45baeb6405568afd88b6/MANIFEST.in#L1) and [sphinx-py3doc-enhanced-theme](https://github.com/ionelmc/sphinx-py3doc-enhanced-theme/blob/50bd06442e9e2a5056ba4ab25e3cbb682f719e50/MANIFEST.in#L3) both are grafting `src` (though Ionel Mărieș [specifically argues for it](https://blog.ionelmc.ro/2014/05/25/python-packaging/))?

Comment (kratsg) [reply MANIFEST.in]: Reading the docs: https://docs.python.org/2/distutils/sourcedist.html

I guess the phrasing is slightly ambiguous. If you want to include additional (non-default) files, then you use that.. but it looks like it's becoming a template to generate the MANIFEST so perhaps this is more correct.

File: docs/schemas (hunk 3 above)
Comment (kratsg) [reply docs/schemas]: Oh, hrmm... Maybe this PR looks awkward. Because it's shown as a _new file_ created.

Comment (matthewfeickert) [reply docs/schemas]: > Maybe this PR looks awkward. Because it's shown as a _new file_ created.

I agree. I was very confused by this behavior as well, but when you [view the file history it shows your previous versions](https://github.com/diana-hep/pyhf/commits/3d16b22be1f6f0a0e5f5d14271476a004ce5d7f0/docs/schemas), even though it is being shown as being "created".

Comment (kratsg) [reply docs/schemas]: Most likely because of how GH treats this kind of file.

File: docs/conf.py [hunk 4]
```diff
@@ -19,7 +19,7 @@
 import os
 import sys
 
-sys.path.insert(0, os.path.abspath('..'))
+sys.path.insert(0, os.path.abspath('../src'))
```
Comment (matthewfeickert) [inline docs/conf.py]: @kratsg Good now?

Comment (matthewfeickert) [reply docs/conf.py]: Looks like it. I pulled this into a fresh Docker container, installed `master`, switched here, built the docs and seems the schema is in good shape too

```
~/data/pyhf/docs$ find _build/html/ -iname "defs.json"
_build/html/schemas/1.0.0/defs.json
```

Thanks very much for catching this in the first place!

**Comment by matthewfeickert** [comment]
@lukasheinrich @kratsg If you two are both okay to ignore the CodeFactor complaints on this, then I think this is ready for review.

File: MANIFEST.in (hunk 1 above)
Comment (matthewfeickert) [reply MANIFEST.in]: Yeah, as you said (from [the 3.7 docs](https://docs.python.org/3.7/distutils/sourcedist.html#specifying-the-files-to-distribute))

>  The typical way to do this is to write a manifest template, called `MANIFEST.in` by default. The manifest template is just a list of instructions for how to generate your manifest file, `MANIFEST`, which is the exact list of files to include in your source distribution. The **sdist** command processes this template and generates a manifest based on its instructions and what it finds in the filesystem.

so it seems that this would be an easy way to be quite sure that everything under `src/` gets packaged up by `sdist`.

**Comment by matthewfeickert** [comment]
CodeFactor is now placated into passing, thanks to @kratsg understanding and then showing that the `setup.py` shebang was causing CodeFactor to run in Python 2 and handle imports differently.
